cpp/src/io/csv/csv_writer.cu

```diff
@@ -323,44 +327,65 @@ gdf_error write_csv(csv_write_arg* args)
         //
         // This is essentially an exclusive-scan (prefix-sum) across columns.
         // Moving left-to-right, add up each column and carry each value to the next column.
-        // Looks like we could transpose the matrix, scan it, and then untranspose it.
-        // Should be able to parallelize the math for this -- will look at prefix-sum algorithms.
+        // We do this in 3 kernel calls below.
         //
-        std::vector<char> buffer(memsize+1);
-        std::vector<size_t> string_locations(rows*count); // all the memory pointers for each column
+        rmm::device_vector<size_t> string_locations(rows*count); // all the memory pointers for each column
         string_locations[0] = 0; // first one is always 0
-        // compute offsets as described above into locations matrix
-        size_t offset = 0;
-        for( gdf_size_type jdx=0; jdx < rows; ++jdx )
+        size_t* d_string_locations = string_locations.data().get();
         {
-            // add up column values for each row
-            // this is essentially an exclusive-scan across columns
-            string_locations[jdx] = (size_t)(buffer.data() + offset); // initialize first item
+            rmm::device_vector<size_t> sums(rows);
+            size_t* d_sums = sums.data().get();
+            // do vertical scan -- add up lengths in each column
+            thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), rows,
+                [d_sums, d_string_lengths, rows, count, d_string_locations] __device__ (unsigned int idx) {
+                    size_t sum = 0;
+                    for( int column=0; column < count; ++column )
+                    {
+                        int columnar_index = (column*rows) + idx;
+                        d_string_locations[columnar_index] = sum;
+                        sum += d_string_lengths[columnar_index];
+                    }
+                    d_sums[idx] = sum;
+                });
+            // scan these intermediate sums
+            thrust::exclusive_scan(execpol->on(0), sums.begin(), sums.end(), sums.begin());
+            // finish off the full scan by adding the intermediate sums to each element
+            thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), rows,
```

Comment:
Parallelize over all elements rather than using an inner loop -- more parallel and code is clearer.

```
thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), rows*count,
                            [d_sums] __device__ (unsigned int idx) { d_string_locations[idx] += d_sums[idx % rows]; });

Reply:
Done. This is much cleaner.